### PR TITLE
[compilers] support for CUDA compilation via nvcc

### PR DESCRIPTION
**Pull request by robertmaynard**
To get sccache to support the nvcc compiler I had to make the following changes to how compilers work. As this is my first time modifying the sccache code base, I am happy to refactor these changes to better suite what the project is looking for.

1. The compilation flag (`-c`) is not hard-coded but extracted from the compilation. This is necessary as the nvcc compiler uses different compilation flags for different types of compilation such as: `-c` ( whole compilation ), `-dc` (separable compilation ), `-ptx` ( ptx generation ). This merge request adds support for all three above compilation types, and follow work can be done (if needed) to also support `-dw`, `-cubin`, and `-fatbin`.

2. Dependency related flags ( -MT, -MF, -MD ) are stored into a separate vector and not the pre-processor flags. This was done for the following reasons:

 - The dependency information only needs to be generated once for each compilation, but since it was stored in `preprocessor_args` it would cause generation to occur both in the `preprocess` pass, and `generate_compile_commands`. This refactor now makes this generation occur during the `preprocess` pass only.

- Removal of the hard-coded `-MT` and fixed a bug related to `-MQ`. When `need_explicit_dep_target` is detected the `gcc` compiler will hard-code adding `-MT` to the preprocessor line, this can be wrong as the user initial flag set included `-MQ` and that should be used. I refactored this logic to handle picking up either `-MT` or `-MQ` from the user, and if neither is provided falling back to adding `-MT`.

- The nvcc compiler doesn't support generation of dependency information when requesting pre-processed output ( `-E` ). This was the driving need for the separation of dependency flags so I could safely invoke the compiler with `-E`. To get everything to work correctly the nvcc `preprocess` pass does 2 invocations of the compiler to generate dep info and pre-processed output.

This set of changes has been in use on the VTK-m project ( https://gitlab.kitware.com/vtk/vtk-m/- ) for over 10 days without issues. In this time sccache has handled ~400k compilation units, with CUDA counting for ~100k of those.

### CODE REVIEW

**Comment by luser** [comment]
I'm not an active maintainer anymore but I just wanted to drop in to say that this is very cool and I'm glad that you were able to make it work for your purposes! Ever since I ported sccache from Python to Rust years ago I've been really happy about how many contributions we've had from users scratching their own itches. I hope you found the code base reasonable to work with!

**Comment by robertmaynard** [comment]
> I hope you found the code base reasonable to work with!

This was my first time reading and writing rust code, and the sccache code was very easy to understand and extend

File: src/compiler/nvcc.rs [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+    }
+
+    #[test]
+    fn test_parse_arguments_simple_c() {
+        let a = parses!("-c", "foo.c", "-o", "foo.o");
+        assert_eq!(Some("foo.c"), a.input.to_str());
+        assert_eq!(Language::C, a.language);
+        assert_map_contains!(a.outputs, ("obj", PathBuf::from("foo.o")));
+        //TODO: fix assert_map_contains to assert no extra keys!
+        assert_eq!(1, a.outputs.len());
```
Comment (mathstuf) [inline src/compiler/nvcc.rs:233]: See #731 (I'm fine with rebasing afterwards).

**Comment by froydnj** [comment]
Echoing @luser's comments: this is very cool, thank you for doing this!

My time is spoken for in other ways this week, but I can try to make some room for reviewing this, or perhaps @glandium has thoughts.  Otherwise, I will look at this early next week.

File: src/compiler/nvcc.rs [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+                } else {
+                    transformed_deps.push(item.clone());
+                }
+            }
+            dep_cmd.args(&transformed_deps)
+                   .env_clear()
+                   .envs(env_vars.iter().map(|&(ref k, ref v)| (k, v)))
+                   .current_dir(cwd);
+
+            if log_enabled!(Trace) {
```
[the remark this answers is not in the document]
Comment (robertmaynard) [reply src/compiler/nvcc.rs:107]: > But this pattern seems to be used more widely, so something better fit for a followup more likely.

Yeah, I was following the previous pattern for the preprocessor invocation which used this pattern

**Pull request merged**